docu xsdeploy scenario (#840)
SAP/jenkins-library: documentation/docs/scenarios/xsa-deploy/Readme.md

## Prerequisites

* [Docker environment](https://docs.docker.com/get-started/)
* All artifacts refereneced during the build are available either on Service Market Place or via public repositories
* You have set up Project “Piper”. See [guided tour](https://sap.github.io/jenkins-library/guidedtour/).
* Docker image for xs deployment available. Due to legal reasons there is no pre-build docker image. How to create the docker image is explained [here](https://github.com/SAP/devops-docker-images/tree/master/xs-cli).
### Project Prerequisites

This scenario requires additional files in your project and in the execution environment on your Jenkins instance.

On the project level, provide and adjust the following template:

| File Name | Description | Position |
|-----|-----|-----|
| [`mta.yaml`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/xsa-deploy/files/mta.yaml) | This file controls the behavior of the MTA toolset. | Place the `mta.yaml` file in your application root folder and adjust the values in brackets with your data. |

> 

Reviewer: As an idea I would just refer to the XS advanced developer quick start guide. https://help.sap.com/viewer/400066065a1b46cf91df0ab436404ddc/2.0.04/en-US/7f681c32c2a34735ad85e4ab403f8c26.html

It describes the entire project structure and the needed files.

> # Build and Deploy SAPUI5 or SAP Fiori Applications on HANA XS with Jenkins

Reviewer 2: ```suggestion
# Build and Deploy SAP Fiori Applications on SAP HANA XS Advanced
```

Reviewer 2: ```suggestion
# Build and Deploy SAP Fiori Applications on SAP HANA XS Advanced
```

On the project level, provide and adjust the following template:

| File Name | Description | Position |
|-----|-----|-----|
| [`mta.yaml`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/xsa-deploy/files/mta.yaml) | This file controls the behavior of the MTA toolset. | Place the `mta.yaml` file in your application root folder and adjust the values in brackets with your data. |

## Context

This scenario combines various different steps to create a complete pipeline.

In this scenario, we want to show how to build a Multitarget Application (MTA) and deploy the build result into an on-prem SAP HANA XS advances system. This document comprises the [mtaBuild](https://sap.github.io/jenkins-library/steps/mtaBuild/) and the [xsDeploy](https://sap.github.io/jenkins-library/steps/xsDeploy/) steps.

![This pipeline in Jenkins Blue Ocean](images/pipeline.jpg)
###### Screenshot: Build and Deploy Process in Jenkins

## Example

### Jenkinsfile

Following the convention for pipeline definitions, use a `Jenkinsfile` which resides in the root directory of your development sources.

> TODO: here we describe the straight-forward case. There is also a blue-green mode. How should we handle this

Reviewer 2: Straight forward first.

Author: Agreed and this is what I did. But finally I think we should also describe the other case. This here is the reminder for that.

```groovy

> @Library('piper-library-os') _

Reviewer 2: Any negative impact when we change this to `piper-lib-os`?

Author: done

> # Build and Deploy SAPUI5 or SAP Fiori Applications on SAP Cloud Platform with Jenkins

Author: done

Build an application based on SAPUI5 or SAP Fiori with Jenkins and deploy the build result into an SAP Cloud Platform account in the Neo environment.

## Prerequisites

> 

Author: @1084565 @SarahNoack  I guess this comment refers to my `TODO` in the code :

> do we have a general description how to setup docker.

Is such a description available so that we can refer to it?

> 

Reviewer 2: We should not describe the docker setup. But maybe I'm missing some aspects.
How would it be different from https://docs.docker.com/get-started/ ?

> 

Author: It is not about describing the docker setup here, that would be clearly the wrong place. But I'm looking for a page describing the docker setup so that I can refer to that from here. Above ypu proivded a like to an external page, that we can use, of course.

* You have installed the Java Runtime Environment 8.
* You have installed Jenkins 2.60.3 or higher.

> * You have set up Project “Piper”. See [README](https://github.com/SAP/jenkins-library/blob/master/README.md).

Author: I changed the link so that it points to the `guided tour`now. I guess this is what you meant.

* You have installed the Multi-Target Application (MTA) Archive Builder 1.0.6 or newer. See [SAP Development Tools](https://tools.hana.ondemand.com/#cloud). TODO: maybe we should refer to the docker image
* You have installed Node.js including node and npm. See [Node.js](https://nodejs.org/en/download/). TODO: docker ?

> 

Author: In case they are available on SMP: should we also explain how the artifacts needs to be provided in the build environment of the customer. Here I see basically two possibilities:

1.) local m2 repo
2.) nexus on customer side

Ad 1.) that was state of the art before docker. With docker we have to mount the local repo in each docker container performing a java build.

Ad 2.) Nexus can be setup based on a docker image.  Needs to be included in `settings.xml`.

Regardless which option we recommand we should also explain how the artifacts can be provided there (e.g. `install:install-file`, `deploy:deploy-file`. For nexus there is also a bulk upload available (for nexus OS a corresponding plugin needs to be installed first). Bulk upload into a local repo can be performed by expanding a suitable `tar.gz`

> 

Reviewer 2: We will have Java dependencies on maven by end of december, so that is the default case.

For the non-default (SMP) my suggestion would be 

> In case you require dependencies that are not available on maven central, you will need to download them available at build-time, preferably by deploying them to your local artifact repository (e.g. nexus).

### Project Prerequisites

This scenario requires additional files in your project and in the execution environment on your Jenkins instance.

On the project level, provide and adjust the following template:

| File Name | Description | Position |
|-----|-----|-----|

> | [`.npmrc`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/.npmrc) | This file contains a reference to the SAP NPM registry (`@sap:registry https://npm.sap.com`), which is required to fetch the dependencies required to build the application. | Place the `.npmrc` file in the root directory of your project. |

Reviewer 3: ```suggestion
| [`.npmrc`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/.npmrc) | This file contains a reference to the SAP NPM registry (`@sap:registry https://npm.sap.com`), which you need to fetch the dependencies required to build the application. | Place the `.npmrc` file in the root directory of your project. |
```

Author: @SarahNoack your comment is outdated. We do not describe technology specific files here again, we rely solely on `mtaBuild` and `xsDeploy`. What we build through `mtaBuild` is in the meantime considered to be out of scope.

> | [`.npmrc`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/.npmrc) | This file contains a reference to the SAP NPM registry (`@sap:registry https://npm.sap.com`), which is required to fetch the dependencies required to build the application. | Place the `.npmrc` file in the root directory of your project. |

Author: Any technology specific file has been dropped in the meantime.

| [`mta.yaml`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/mta.yaml) | This file controls the behavior of the MTA toolset. | Place the `mta.yaml` file in your application root folder and adjust the values in brackets with your data. |
| [`package.json`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/package.json) | This file lists the required development dependencies for the build. | Add the content of the `package.json` file to your existing `package.json` file. |

> | [`Gruntfile.js`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/Gruntfile.js) | This file controls the grunt build. By default the tasks `clean`, `build`, and `lint` are executed. | Place the `Gruntfile.js` in the root directory of your project. |

Reviewer 3: ```suggestion
| [`Gruntfile.js`](https://github.com/SAP/jenkins-library/blob/master/documentation/docs/scenarios/ui5-sap-cp/files/Gruntfile.js) | This file controls the Grunt build. By default, the tasks `clean`, `build`, and `lint` are executed. | Place the `Gruntfile.js` in the root directory of your project. |
```

Author: @SarahNoack your comment is outdated. We do not describe technology specific files here again, we rely solely on `mtaBuild` and `xsDeploy`. What we build through `mtaBuild` is in the meantime considered to be out of scope.

## Context

This scenario combines various different steps to create a complete pipeline.

> In this scenario, we want to show how to build an application based on SAPUI5 or SAP Fiori by using the multi-target application (MTA) concept and how to deploy the build result into an SAP Cloud Platform account in the Neo environment. This document comprises the [mtaBuild](https://sap.github.io/jenkins-library/steps/mtaBuild/) and the [neoDeploy](https://sap.github.io/jenkins-library/steps/neoDeploy/) steps.

Reviewer 3: ```suggestion
In this scenario, we want to show how to build an application based on SAPUI5 or SAP Fiori by using the multitarget application (MTA) concept and how to deploy the build result into an SAP Cloud Platform account in the Neo environment. This document comprises the [mtaBuild](https://sap.github.io/jenkins-library/steps/mtaBuild/) and [neoDeploy](https://sap.github.io/jenkins-library/steps/neoDeploy/) steps.
```

Author: @SarahNoack In the meantime we have:

> In this scenario, we want to show how to build a Multitarget Application (MTA) and deploy  [...].

That is close to your proposal (but it is upper-case). Should we switch to lower-case?

Reviewer 2: @SarahNoack I don't find MTA in the naming guidelines _at all_. How can that be?

![This pipeline in Jenkins Blue Ocean](images/pipeline.jpg)
###### Screenshot: Build and Deploy Process in Jenkins

## Example

### Jenkinsfile

> Following the convention for pipeline definitions, use a `Jenkinsfile` which resides in the root directory of your development sources.

Reviewer 3: ```suggestion
Following the convention for pipeline definitions, use a `Jenkinsfile`, which resides in the root directory of your development sources.
```

```groovy
@Library('piper-lib-os') _

node() {
    stage('prepare') {
        deleteDir
        checkout scm
        setupCommonPipelineEnvironment script: this
    } 
    stage('build') {

>         mtaBuild: script: this // mode XSA should be provided in .pipeline/config.yml

Reviewer 2: ```suggestion
        mtaBuild: script: this // mtaBuild.buildTarget should be set to "XSA" in .pipeline/config.yml
```

    }

    stage('deploy') {
        xsDeploy scipt: this
    }
}
```

### Configuration (`.pipeline/config.yml`)

This is a basic configuration example, which is also located in the sources of the project.

```yaml
steps:
  mtaBuild:
    buildTarget: 'XSA'

>     mtaJarLocation: '/opt/sap/mta.jar' // not sure if this is needed (docker)

Reviewer 2: ja, lets have docker here

  xsDeploy:
    credentialsId: 'xsa'
    mode: 'DEPLOY'
    org: 'myOrg'
    space: 'mySpace'
```

#### Configuration for the MTA Build

| Parameter        | Description    |
| -----------------|----------------|

> | `buildTarget`    | The target platform to which the mtar can be deployed. In this case we need  `XSA` |

Reviewer 3: ```suggestion
| `buildTarget`    | The target platform to which the mtar can be deployed. In this case, the target platform is  `XSA`. |
```

#### Configuration for the Deployment to XSA 

| Parameter          | Description |
| -------------------|-------------|

> | `credentialsId` | The Jenkins credentials that contain the user and password which are used for the deployment on SAP Cloud Platform.|

Reviewer 3: ```suggestion
| `credentialsId` | The Jenkins credentials that contain user and password required for the deployment on SAP Cloud Platform.|
```
